zerver/lib/streams.py

```diff
@@ -131,26 +132,34 @@ def create_streams_if_needed(
     stream_dicts: List[StreamDict],
     acting_user: Optional[UserProfile]=None
 ) -> Tuple[List[Stream], List[Stream]]:
-    """Note that stream_dict["name"] is assumed to already be stripped of
+    """Note that stream_dict["stream"] is name and assumed to already be stripped of
     whitespace"""
     added_streams: List[Stream] = []
     existing_streams: List[Stream] = []
     for stream_dict in stream_dicts:
-        stream, created = create_stream_if_needed(
-            realm,
-            stream_dict["name"],
-            invite_only=stream_dict.get("invite_only", False),
-            stream_post_policy=stream_dict.get("stream_post_policy", Stream.STREAM_POST_POLICY_EVERYONE),
-            history_public_to_subscribers=stream_dict.get("history_public_to_subscribers"),
-            stream_description=stream_dict.get("description", ""),
-            message_retention_days=stream_dict.get("message_retention_days", None),
-            acting_user=acting_user
-        )
-
-        if created:
-            added_streams.append(stream)
+        # Rechecking that stream_dict["stream"] is `name` (string) to avoid passing `id` (int) to it.
+        if isinstance(stream_dict["stream"], str):
+            stream, created = create_stream_if_needed(
+                realm,
+                str(stream_dict["stream"]),
+                invite_only=stream_dict.get("invite_only", False),
+                stream_post_policy=stream_dict.get("stream_post_policy", Stream.STREAM_POST_POLICY_EVERYONE),
+                history_public_to_subscribers=stream_dict.get("history_public_to_subscribers"),
+                stream_description=stream_dict.get("description", ""),
+                message_retention_days=stream_dict.get("message_retention_days", None),
+                acting_user=acting_user
+            )
+
+            if created:
+                added_streams.append(stream)
+            else:
+                existing_streams.append(stream)
         else:
-            existing_streams.append(stream)
+            # Raising error if unknown stream id is passed.
+            try:
+                stream = get_stream_by_id_in_realm(stream_dict["stream"], realm)
+            except Stream.DoesNotExist:
+                raise JsonableError(_("No stream with id '{}' found.").format(stream_dict["stream"]))
```

Comment:
Rather than raising the error here, I think it would be better to enforce that `autocreate=True` is never passed to `list_to_streams` when using `stream_id` as identifier. (Maybe raise an error in `list_to_streams` itself if this is not followed?)

Reply:
Yes @abhijeetbodas2001,  I agree with you it will be better this way.